source/fuzz/transformation_add_synonym.cpp

```diff
@@ -0,0 +1,156 @@
+    protobufs::TransformationAddSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationAddSynonym::TransformationAddSynonym(
+    uint32_t result_id, const protobufs::InstructionDescriptor& insert_before,
+    const protobufs::Instruction& synonymous_instruction) {
+  message_.set_result_id(result_id);
+  *message_.mutable_insert_before() = insert_before;
+  *message_.mutable_synonymous_instruction() = synonymous_instruction;
+}
+
+bool TransformationAddSynonym::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // Check that |message_.synonym_id| is valid.
+  auto* synonym = ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  if (!synonym) {
+    return false;
+  }
+
+  // Check that |insert_before| is valid.
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+  if (!insert_before_inst) {
+    return false;
+  }
+
+  // Check that we can insert |message._synonymous_instruction| before
+  // |message_.insert_before| instruction.
+  auto opcode = static_cast<SpvOp>(message_.synonymous_instruction().opcode());
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(opcode,
+                                                    insert_before_inst)) {
+    return false;
+  }
+
+  // Check that synonymous instruction has fresh id.
+  if (!fuzzerutil::IsFreshId(ir_context,
+                             message_.synonymous_instruction().result_id())) {
+    return false;
+  }
+
+  // Domination rules must be satisfied.
+  if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+          ir_context, insert_before_inst, message_.result_id())) {
+    return false;
+  }
+
+  // Check that new synonymous instruction is valid.
+  switch (opcode) {
+    case SpvOpIAdd:
+    case SpvOpIMul:
+    case SpvOpFAdd:
+    case SpvOpFMul:
+    case SpvOpLogicalOr:
+    case SpvOpLogicalAnd: {
+      if (message_.synonymous_instruction().input_operand_size() != 2) {
+        return false;
+      }
+
+      const auto& lhs = message_.synonymous_instruction().input_operand(0);
+      const auto& rhs = message_.synonymous_instruction().input_operand(1);
+      if (lhs.operand_type() != SPV_OPERAND_TYPE_ID ||
+          rhs.operand_type() != SPV_OPERAND_TYPE_ID) {
+        return false;
+      }
+
+      auto lhs_type_id = fuzzerutil::GetTypeId(ir_context, lhs.operand_data(0));
+      auto rhs_type_id = fuzzerutil::GetTypeId(ir_context, rhs.operand_data(0));
+      if (lhs_type_id != rhs_type_id ||
+          lhs_type_id != message_.synonymous_instruction().result_type_id()) {
+        return false;
+      }
+
+      const auto* type = ir_context->get_type_mgr()->GetType(lhs_type_id);
+      if (!type) {
+        return false;
+      }
+
+      switch (opcode) {
+        case SpvOpIAdd:
+        case SpvOpIMul:
+          return type->AsInteger() ||
+                 (type->AsVector() &&
+                  type->AsVector()->element_type()->AsInteger());
```

Comment:
The problem I see with this transformation is: just because the instruction has an argument of the right type, it doesn't follow that the transformation will be semantics-preserting.

For instance, for the case of addition we need to know that the thing being added is zero; for multiplication we need to know that the thing being used as a multiplicand is 1.

It doesn't seem that your transformation checks this, and it would be really cumbersome to make it do so.

Reply:
Fixed in ecf877b.